upstream(indexer): Rework config and cli arguments

tomxey opened the pull request:
# Description of change

This patch reworks a number of the indexer configs and cli arguments to try and centralize where the configuration for the indexer is loaded. This also cleans up the cli arguments into well defined subcommands instead of using boolean flags for deciding if a jsonrpc or indexer writer service should be started.

The old CLI is still kept for backward compatibility with an intention to remove it in the future.

## Links to any relevant issues

fixes: #6201
fixes: https://github.com/iotaledger/iota/issues/7119

## Type of change

Rafactoring

## How the change has been tested

`cargo ci-clippy`
`cargo test --profile simulator --package iota-indexer --test rpc-tests --all-features`

tested the `pg-services-local` docker setup, it starts and indexer-sync is syncing data, rpc worker and analytical worker also start successfully


## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes

### Release Notes

<!--
Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates.
-->

- [ ] Protocol:
- [ ] Nodes (Validators and Full nodes):
- [x] Indexer:
 New Indexer CLI was introduced. :warning: The old CLI is still kept for backward compatibility but will be removed in the future so all users are encouraged to start using new CLI when possible. 
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

tomxey committed:
refactor(indexer): Rework config and cli arguments

tomxey committed:
Fix param name for tests

sergiupopescu199 committed:
Merge branch 'develop' into sc-platform/refactor-indexer-cli

# Conflicts:
#	crates/iota-indexer/src/db.rs
#	crates/iota-indexer/src/lib.rs

kodemartin commented on crates/iota-indexer/src/main.rs:
```diff
@@ -19,26 +58,180 @@ async fn main() -> Result<(), IndexerError> {
+                    ingestion_config: IngestionConfig {
+                        sources: IngestionSources {
+                            data_ingestion_path: old_conf.data_ingestion_path,
+                            remote_store_url: old_conf.remote_store_url.map(|url| {
+                                url.parse().expect("Remote Store URL should be correct")
+                            }),
+                            rpc_client_url: Some(rpc_client_url_parsed),
+                        },
+                        checkpoint_download_queue_size: download_queue_size,
+                        checkpoint_download_timeout: ingestion_reader_timeout_secs,
+                        checkpoint_download_queue_size_bytes: data_limit,
+                    },
+                    snapshot_config: SnapshotLagConfig {
+                        snapshot_min_lag,
+                        sleep_duration: SnapshotLagConfig::DEFAULT_SLEEP_DURATION_SEC,
+                    },
+                    pruning_options: PruningOptions {
+                        epochs_to_keep: std::env::var("EPOCHS_TO_KEEP")
+                            .map(|s| s.parse::<u64>().ok())
+                            .unwrap_or_else(|_e| None),
+                    },
+                }
+            };
+
+            iota_indexer::config::IndexerConfig {
+                database_url: db_url
+                    .map_err(|e| {
+                        IndexerError::PgPoolConnection(format!(
+                            "Failed parsing database url with error {e:?}"
+                        ))
+                    })?
+                    .expose_secret()
+                    .parse()
+                    .expect("Database URL should be correct"),
+                connection_pool_config: pool_config_from_env(),
+                metrics_address,
+                command,
+            }
+        }
+    };
```
Failing to parse the new CLI may mean that `--help` is requested, or any other user-input error that might require feedback with printing help.

So falling back to the deprecated should be avoided. 

We should instead simplify things and first try to parse the deprecated config and on error fallback to the CLI allowing any error to be handled by `clap`. Something like

```rust
    let opts = match DeprecatedIndexerConfig::try_parse() {
        Ok(opts) => opts.try_into()?,
        Err(_) => IndexerConfig::parse(),
    };
```
with proper modifications to account for `version`.

This enforces the following behaviour:

* We support successful usage of the deprecated interface,
* Otherwise, we direct to usage of the new interface

kodemartin replied:
As implied by the previous comment it makes sense to implement `TryFrom<DeprecatedIndexerConfig> for IndexerConfig`.

kodemartin commented on crates/iota-indexer/src/main.rs:
```diff
@@ -19,26 +58,180 @@ async fn main() -> Result<(), IndexerError> {
+                    rpc_client_url: old_conf.rpc_client_url,
+                })
+            } else {
+                Command::Indexer {
+                    ingestion_config: IngestionConfig {
+                        sources: IngestionSources {
+                            data_ingestion_path: old_conf.data_ingestion_path,
+                            remote_store_url: old_conf.remote_store_url.map(|url| {
+                                url.parse().expect("Remote Store URL should be correct")
+                            }),
+                            rpc_client_url: Some(rpc_client_url_parsed),
+                        },
+                        checkpoint_download_queue_size: download_queue_size,
+                        checkpoint_download_timeout: ingestion_reader_timeout_secs,
+                        checkpoint_download_queue_size_bytes: data_limit,
+                    },
+                    snapshot_config: SnapshotLagConfig {
+                        snapshot_min_lag,
+                        sleep_duration: SnapshotLagConfig::DEFAULT_SLEEP_DURATION_SEC,
+                    },
+                    pruning_options: PruningOptions {
+                        epochs_to_keep: std::env::var("EPOCHS_TO_KEEP")
+                            .map(|s| s.parse::<u64>().ok())
+                            .unwrap_or_else(|_e| None),
+                    },
+                }
```
In this case it should fail. Currently if users don't define the kind of worker it leads to a noop.

kodemartin commented on crates/iota-indexer/src/main.rs:
```diff
@@ -19,26 +58,180 @@ async fn main() -> Result<(), IndexerError> {
+
+            let command = if old_conf.analytical_worker {
+                Command::AnalyticalWorker
+            } else if old_conf.reset_db {
+                Command::ResetDatabase { force: true }
```
This is not mapping to the existing behaviour where resetting the db is not a standalone command.

I believe that we should just nest `reset-db` as an option to the `indexer` command in the new CLI as well.

kodemartin commented on crates/iota-indexer/src/config.rs:
since we are adding a new module it seems proper to nest the deprecated config here under a `deprecated` module.

kodemartin replied:
Nit: Move to a dedicated file:

```
config/
  mod.rs # 
  deprecated.rs
```

tomxey replied:
Since we are about to remove it in the future I would prefer to leave it in the same file, to not shuffle files back and forth when we will eventually delete it.

kodemartin commented on crates/iota-indexer/src/config.rs:
```diff
@@ -0,0 +1,289 @@
+    about = "An off-fullnode service serving data from IOTA protocol"
+)]
+pub struct IndexerConfig {
+    #[clap(long, alias = "db-url")]
```
The `clap` attribute seems deprecated. Please use `arg` instead and other attributes according to the latest [derive reference](https://docs.rs/clap/latest/clap/_derive/index.html)

kodemartin commented on crates/iota-indexer/src/config.rs:
```diff
@@ -0,0 +1,289 @@
+        #[clap(long)]
+        force: bool,
+    },
+    AnalyticalWorker,
```
Nit: We could add a `HelpDeprecated` command to print the deprecated help. 

This would require to make the `database-url` an `Option` and validate its presence for other commands after parsing.

kodemartin commented on crates/iota-indexer/src/config.rs:
```diff
@@ -0,0 +1,289 @@
+        pruning_options: PruningOptions,
+    },
+    JsonRpcService(JsonRpcConfig),
+    ResetDatabase {
```
I think it is better to nest this under `Indexer`

kodemartin reviewed (changes requested):
Hey @tomxey, I think a few revisions are in order. Here's a summary

* Simplify logic by supporting only successful usage of deprecated CLI
* Use new cli as a fallback
* Provide optionally  a `help-deprecated` subcommand
* Implement `TryFrom<DeprecatedConfig> for IndexerConfig`
* Nest the deprecated config under `config::deprecated`
* Use latest clap derive reference
* Nest `--reset-db` under `Indexer` command

See more specific comments as well.

Plus please update the `README` with reference to both deprecated usage and the new usage.

tomxey committed:
Move old config to deprecated module, make reset db part of sync worker, fail if worker type is not specified

tomxey committed:
Try to parse with old CLI initially, fall back to the new one
Move conversion logic to TryFrom
Add help-deprecated command
Document usages of old and new cli

Thoralf-M commented on crates/iota-indexer/src/main.rs:
```diff
@@ -19,26 +32,83 @@ async fn main() -> Result<(), IndexerError> {
+        )?;
+        spawn_connection_pool_metric_collector(indexer_metrics.clone(), connection_pool.clone());
+
+        match opts.command {
```
Can we change this so one can run Indexer, JsonRpcServcice and AnalyticalWorker at once?
I guess for full control/to have it more resilient it makes sense to run them separated, but for testing it would be nice to have the option to start them all in a single command (the `iota start` command does this as well, but here ofc it should be without all the other stuff)

tomxey replied:
I think we could add a separate command that runs all 3 worker types at once, but I would prefer to tackle it in a separate issue.
Could you create one if it works for you?

Thoralf-M replied:
Sure, https://github.com/iotaledger/iota/issues/7318

kodemartin commented on crates/iota-indexer/src/config.rs:
```diff
@@ -0,0 +1,556 @@
+    },
+    JsonRpcService(JsonRpcConfig),
+    AnalyticalWorker,
+    HelpDeprecated,
```
Nit: Add docstring to show while calling `--help`
```suggestion
    /// Print help for the deprecated interface.
    HelpDeprecated,
```

kodemartin commented on crates/iota-indexer/src/main.rs:
```diff
@@ -19,26 +32,83 @@ async fn main() -> Result<(), IndexerError> {
-            "{}:{}",
-            indexer_config.client_metric_host, indexer_config.client_metric_port
+    let old_conf = OldIndexerConfig::command()
+        .version(VERSION)
```
We don't actually need the version for the old config. `--version` is ensured and can be only used by the new interface.

So this can be simplified, using higher-level `clap` methods.

kodemartin commented on crates/iota-indexer/src/main.rs:
```diff
@@ -19,26 +32,83 @@ async fn main() -> Result<(), IndexerError> {
+        OldIndexerConfig::command().print_help().map_err(|e| {
+            IndexerError::Generic(format!("Failed printing deprecated CLI help: {e}"))
+        })?;
+    } else {
```
Nit: We can save one level of indentation by simplifying to:

```rust
if let Command::HelpDeprecated = opts.command {
    OldIndexerConfig::command().print_help().map_err(|e| {
        IndexerError::Generic(format!("Failed printing deprecated CLI help: {e}"))
    })?;
    return Ok(());
}

let connection_pool = new_connection_pool(
    opts.database_url
        .ok_or(IndexerError::InvalidArgument(
            "--database-url argument is mandatory for this command".into(),
        ))?
        .as_str(),
    &opts.connection_pool_config,
)?;
spawn_connection_pool_metric_collector(indexer_metrics.clone(), connection_pool.clone());

match opts.command {
    Command::Indexer { ingestion_config, snapshot_config, pruning_options, reset_db } => {
        // ... logic for Indexer ...
    }
    Command::JsonRpcService(json_rpc_config) => {
        // ... logic for JsonRpcService ...
    }
    Command::AnalyticalWorker => {
        // ... logic for AnalyticalWorker ...
    }
    Command::HelpDeprecated => unreachable!("Already handled above"),
}
```

kodemartin reviewed (approved):
Looks good! Left a few nitpicks.

tomxey committed:
Update crates/iota-indexer/src/config.rs

Co-authored-by: Konstantinos Demartinos <konstantinos.demartinos@iota.org>

tomxey committed:
Simplifi parsing the old config, reduce level of indentation

TheMrAI commented on crates/iota-indexer/src/db.rs:
```diff
@@ -17,17 +18,28 @@ use crate::errors::IndexerError;
+    #[arg(env = "DB_CONNECTION_TIMEOUT")]
     pub connection_timeout: Duration,
+    #[arg(long, value_parser = parse_duration, default_value = "3600")]
+    #[arg(env = "DB_STATEMENT_TIMEOUT")]
```
Noice!

TheMrAI reviewed (approved):
LGTM, some nice refactors.

tomxey committed:
Merge branch 'develop' into sc-platform/refactor-indexer-cli
